docs/source/_static/rest-api.yml

```diff
@@ -135,6 +135,27 @@ paths:
           schema:
             type: boolean
           allowEmptyValue: true
+        - name: sort
+          in: query
+          description: |
+            Sort users by this key (new in JupyterHub 5.0).
+            Default: id (opaque but stable internal order).
+          schema:
+            type: string
+            enum:
+              - id
+              - name
+              - last_activity
+        - name: direction
```

Comment:
I'm considering the query parameters naming, they are named very briefly.

`direction` isn't self explanatory, and `sort` was something I asked myself "is this a boolean?" about, but learn from reading more that its a `sort key`.

I'm considering options like `sort_by` instead of `sort` and `sort_order` or `sort_direction`instead of `direction` as the query parameter names. We can leave it as it is as well, but I think self-explanatory naming could be worth more than being brief here.

Reply:
I didn't choose them to be brief, I chose this to be exactly the names and values for the same thing in the GitHub API. I don't know if consistency across familiar APIs is better than trying to be clear and verbose in our own names, though. If I were picking my own, I'd do `sort_by` and `sort_direction` (or boolean `sort_ascending`, but query parameters don't have super clear ways to specify bools). WDYT?

Reply:
I think following a naming convention from an existing well known API (I don't have a preference for which one) is better than coming up with our own. Otherwise it's effectively the N+1 standards problem 😁.

Reply:
I looked around from a search "sorting rest api best practices" and then went onwards to look at examples from some big companies exposing REST API's:

- [Ref 1](https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/#sorting), mentions `sort` and `sort_by`, and examplifies how the sort direction can provided by the value of this parameter as well
- [Ref 2](https://www.atatus.com/blog/rest-api-design-filtering-sorting-and-pagination/#sorting), mentions `sort` and `order`, and examplifies multi column sorting by `sort` bundled to handle order as well with a prepended `-` sign for descending.
- [Ref 3](https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/#h2-ef7a235718c20), mentions `sort` with `+-` prefixes to columns to determine order.
- [Ref 4, twitter](https://developer.twitter.com/en/docs/twitter-ads-api/sorting), `sort_by` including suffix `-asc` or `-desc`
- [Ref 5, some facebook api](https://developers.facebook.com/docs/marketing-api/insights/#sorting), `sort` including `_ascending` or `_descending` suffix.

---

There doesn't seem to be a common naming practice to adhere to. Due to this, I'm currently favor we pick something that is a bit more self explanatory, and possibly also a pattern that allows us to re-use it in multi-column sorting situations even though we won't support it. I think currently lean towards favoring a single query parameter `sort_by` with a use of prefix `+` or `-` (`+` / ascending by default), thinking that we won't need a separator to use such prefixes, while we may need a separator if we suffix `-desc` or similar.

@minrk maybe you can make a call at this point and we'll go with it?

Reply:
I really wish naming things wasn't so hard, which is why I always try to follow someone else's choice instead of starting the naming discussion from scratch. [jsonapi](https://jsonapi.org/format/) uses the `-` prefix for descending and defines no prefix as ascending. The reason not to use `+` prefix is that `+` in a query string actually means a space character. I don't think we should suggest using any characters that could get confused by url escaping.

I do like the possibility of chaining sorts with commas in the future, so I'll go with the jsonapi spec for now (`sort=id` or `sort=-last_activity`, no `+`), though I won't implement `sort=-last_activity,name` multi-level sort here. It does give us a clear way to do it in the future.

I really don't think anyone could be confused by the absence of `_by`, since nobody will see the `sort` parameter without also seeing at least one example of a value, which removes all ambiguity. There definitely does seem to be consensus among sampling that `sort` is expected over `sort_by`, even if there are some exceptions, so I think `sort` is the better choice for the parameter name.

Reply:
Excellent find of referencing jsonapi - I'm also happy that this syntax allows for multi-sorting if we want to implement support for that.